File: 表情键盘oc版/表情键盘OC/表情键盘OC/Emoticon/Model/Emoticon.m
//
//  Emoticon.m
//  表情键盘OC
//
//  Created by 单车 on 2019/10/29.
//  Copyright © 2019 单车. All rights reserved.
//

#import "Emoticon.h"

#define EmojiCodeToSymbol(c) ((((0x808080F0 | (c & 0x3F000) >> 4) | (c & 0xFC0) << 10) | (c & 0x1C000) << 18) | (c & 0x3F) << 24)


@implementation Emoticon
- (instancetype)initWithDict:(NSDictionary *)dict{
    self = [super init];
    if (self) {
        self.code = dict[@"code"];
        self.png = dict[@"png"];
        self.chs = dict[@"chs"];
        
        if (self.code != nil) {
            self.emojiCode = [Emoticon emojiWithStringCode:self.code];
        }
        
        if (self.png != nil) {
            self.pngPath = [NSString stringWithFormat:@"%@/Emoticons.bundle/%@",NSBundle.mainBundle.bundlePath,self.png];
        }
        
    }
    return self;
}

- (instancetype)initWithRemove:(BOOL)isRemove{
    self = [super init];
    if (self) {
        self.isRemove = YES;
    }
    return self;
}

- (instancetype)initWithEmpty:(BOOL)isEmpty{
    self = [super init];
    if (self) {
        self.isEmpty = YES;
    }
    return self;
}

#pragma mark - < emoji 表情 >
/**
 将十六进制的编码转为 emoji 字符串
 
 @param intCode 无符号 32 位整数
 @return 字符串
 */
+ (NSString *)emojiWithIntCode:(unsigned int)intCode{
    unsigned int symbol = EmojiCodeToSymbol(intCode);
    NSString *string = [[NSString alloc]initWithBytes:&symbol length:sizeof(symbol) encoding:NSUTF8StringEncoding];
    if (string == nil) {
        string = [NSString stringWithFormat:@"%C",(unichar)intCode];
    }
    return string;
}
/**
将十六进制的编码转为 emoji 字符串

@param stringCode 十六进制格式的字符串, 例如`0x1f633`
@return 字符串
*/
+ (NSString *)emojiWithStringCode:(NSString *)stringCode
{
    NSScanner *scanner = [[NSScanner alloc] initWithString:stringCode];
    
    unsigned int intCode = 0;
    
    [scanner scanHexInt:&intCode];
    
    return [self emojiWithIntCode:intCode];
}
/**
 返回当前十六进制格式字符串`0x1f633`对应的emoji字符串
 
 @return emoji 字符串
 */
//- (NSString *)emoji
//{
//    return [Emoticon emojiWithStringCode:self];
//}
@end

File: 09-表情键盘备份/09-表情键盘/Emoticon/Model/Emoticon.swift
//
//  Emoticon.swift
//  09-表情键盘
//
//  Created by 单车 on 2019/10/22.
//  Copyright © 2019 单车. All rights reserved.
//

import UIKit

class Emoticon: NSObject {
    //MARK:- 定义属性
    var code : String?  //emoji的code
    
    var png : String?   //普通表情对应的图片名称
    {
        didSet {
            guard let png = png else {
                return
            }
            pngPath = Bundle.main.bundlePath + "/Emoticons.bundle/" + png
        }
    }
    var chs : String?   //普通表情对应的文字
    
    //MARK:- 数据处理
    var pngPath : String?
    var emojiCode : String?
    var isRemove : Bool = false
    var isEmpty : Bool = false

    //MARK:- 自定义构造函数
    init(dict:[String : String] ) {
        super.init()
        
        self.code = dict["code"]
        self.png = dict["png"]
        self.chs = dict["chs"]
        
        if self.code != nil {
            self.emojiCode = self.emojiCode(code: self.code!)
        }
        if self.png != nil {
            self.pngPath = Bundle.main.bundlePath + "/Emoticons.bundle/" + png!
        }
    }
    init(isRemove:Bool){
        self.isRemove = true
    }
    
    init(isEmpty:Bool){
        self.isEmpty = true
    }
}
extension Emoticon{
    func emojiCode(code:String) -> String {
        //1.创建扫描器
        let scanner = Scanner(string: code)
        //2.调用方法，扫描出code中的值
        var value : UInt32 = 0
        scanner.scanHexInt32(&value)

        //3.将value转成字符
        let c = Character(UnicodeScalar(value)!)

        //4.将字符c转成字符串
        return String(c)
    }
}
